### PR TITLE
Netlink API update for:

### PR DESCRIPTION
**Pull request by kwabena-bruni**
- connection batching
- filtering by latest activity
  Also some bug fixes (variables at initialization, etc.)

### CODE REVIEW

File: include/net/tcp_estats_nl.h [hunk 1]
```diff
@@ -1,15 +1,21 @@
 #ifndef _TCP_ESTATS_NL_H_
 #define _TCP_ESTATS_NL_H_
 
+/* The netlink commands that we expect to receive. */
 enum nl_estats_msg_types {
 	TCPE_CMD_LIST_CONNS,
-	TCPE_CMD_READ_ALL,
-	TCPE_CMD_READ_VARS,
+	TCPE_CMD_READ_VARS, /* read vars from a single connection */
```
Comment (dmah42) [inline include/net/tcp_estats_nl.h:8]: i think this will require a userspace update to keep them in sync. Why are these reordered?

File: include/net/tcp_estats.h [hunk 2]
```diff
@@ -58,7 +58,7 @@ enum tcp_estats_softerror_reason {
 	TCP_ESTATS_SOFTERROR_OTHER = 8,
 };
 
-#define TCP_ESTATS_INACTIVE	0
+#define TCP_ESTATS_INACTIVE	2
```
Comment (dmah42) [inline include/net/tcp_estats.h:65]: why this change?

File: include/net/tcp_estats_nl.h (hunk 1 above)
Comment (kwabena-bruni) [inline include/net/tcp_estats_nl.h:8]: You are right - If I remember, there was a discrepancy between your patch
and the official 10g patch - can't remember which I went with but I can
check when I am back at my desk.  Order here doesn't matter and I'm happy
to go back to 10g patch order

File: include/net/tcp_estats.h (hunk 2 above)
Comment (kwabena-bruni) [inline include/net/tcp_estats.h:65]: Change of INACTIVE value was changed to match rfc.  It also has benefit of
Identifying bugs where this variable is not set.

**Comment by rapier1** [comment]
Kevin, I've been looking over the patch. If I'm reading it correctly the connection batching seems straightforward. Mostly minor modifications over what you submitted previously. Let me know if this is incorrect and we're about to start updating userland to reflect the previous submission. 

The filter based on a timestamp doesn't _require_ changes to userland but if we want to make use of it we have to add the timestamp to the nlmsg header. This shouldn't be a big deal if that's the case. I am concerned about performance impacts. Have you quantified this at all? 

Lastly, the read_conns_vars is essentially the same thing that web10g-logger does but in kernel space. Essentially it steps through each of the connections, gets the associated values, and returns them as many messages as required. Am I reading that correctly? Again, this would have to be implemented in userland. If that's the case would you object to implementing it in our userland?

**Comment by kwabena-bruni** [comment]
> > If I'm reading it correctly the connection batching seems
> > straightforward.
> > That's correct.
> > 
> > The filter based on a timestamp doesn't _require_ changes to userland but
> > if we want to make use of it we have to add the timestamp to the nlmsg
> > header. This shouldn't be a big deal if that's the case. I am concerned
> > about performance impacts. Have you quantified this at all?
> > The performance is roughly linear in terms of the number of connections
> > returned via netlink - so there's significant savings in the case where
> > you've got a million idle connections and only 10k or so active ones in the
> > period between samples (i.e. roughly 100x boost in the case of 1 million
> > connections, 10k active)
> > 
> > Lastly, the read_conns_vars is essentially the same thing that
> > web10g-logger does but in kernel space.
> > Yes - and it has the added benefit of making sure that the connection
> > doesn't die between reading the connection spec and reading the variables.
> >  Performance-wise, when reading only a few masked variables (6 in my case),
> > it cut the read time in half for 50k connections - probably mostly due to
> > not needing to read the connection spec twice, but also doing fewer
> > transitions to/from kernel code.
> > 
> > If that's the case would you object to implementing it in our userland?
> > Let me give it a shot this weekend - should be relatively straightforward
> > once I can make sense of all the macros :)

cheers!
- Kevin

**Comment by kwabena-bruni** [comment]
Back up to date - along with a few minor bug fixes.  I think this is now substantially correct and backwards compatible with existing userland... (still working on the "read_all" in my fork of C-based userland)
